iceaxe/__tests__/conftest.py

```diff
@@ -1,84 +1,300 @@
+import logging
+import socket
+import time
+import uuid
+
 import asyncpg
+import docker
 import pytest
 import pytest_asyncio
+from docker.errors import APIError
 
 from iceaxe.base import DBModelMetaclass
 from iceaxe.session import DBConnection
 
+# Configure logging
+logging.basicConfig(level=logging.INFO)
+logger = logging.getLogger(__name__)
+
+
+def get_free_port():
+    """Find a free port on the host machine."""
+    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
+        s.bind(("", 0))
+        return s.getsockname()[1]
+
+
+def stop_containers_using_port(client, port):
+    """Stop any containers using the specified port."""
+    for container in client.containers.list():
+        container_ports = container.attrs.get("HostConfig", {}).get("PortBindings", {})
+        for container_port, bindings in container_ports.items():
+            for binding in bindings:
+                if binding.get("HostPort") == str(port):
+                    logger.info(
+                        f"Stopping container {container.name} that is using port {port}"
+                    )
+                    try:
+                        container.stop()
+                        container.remove()
+                        logger.info(
+                            f"Successfully stopped and removed container {container.name}"
+                        )
+                    except Exception as e:
+                        logger.warning(
+                            f"Failed to stop container {container.name}: {e}"
+                        )
+
+    # Double-check if the port is free now
+    try:
+        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
+            s.bind(("0.0.0.0", port))
+            return True
+    except OSError:
+        return False
+
+
+@pytest.fixture(scope="session")
+def docker_postgres():
+    """
+    Fixture that creates a PostgreSQL container using the Python Docker API.
+    This allows running individual tests without needing Docker Compose.
+    """
+    # Initialize Docker client
+    client = docker.from_env()
+
+    # Generate a unique container name to avoid conflicts
+    container_name = f"postgres-test-{uuid.uuid4().hex[:8]}"
+
+    # First try with the fixed port to match existing test expectations
+    preferred_port = 5438
```

Comment:
If we're switching to a dynamic script approach anyway, I would advise just getting a new free port from the system in a new fixture instead of hard-coding it here.

Stopping containers potentially created outside of this testing harness seems like outside the assumed permission scope for a unit test.